### PR TITLE
[NPUW] Enable mask skipping for fused flash attention 

### PR DESCRIPTION
**Pull request by kkoryun**
### Details:
PR adds logic for working with a attention mask:
 - Skip mask operations (in kernel) for middle tiles (1,3,4,6,7,8), use a mask for the final tiles (0,2,5,9)
image

New subgraph without mask input added.

### Tickets:
 - E*215657

### AI Assistance:
 - *AI assistance used: no / yes*
 - *If yes, summarize how AI was used and what human validation was performed (build/tests/manual checks).*

### CODE REVIEW

File: src/plugins/intel_npu/src/plugin/npuw/attn/attn_subgraph.cpp [hunk 1]
```diff
@@ -39,9 +39,10 @@ struct PyramidRequestSet {
 
 struct HFARequestSet {
     enum TileIdx : std::size_t {
-        REGULAR_TILE = 0,
-        FINAL_TILE = 1,
-        COUNT = 2,
+        REGULAR_TILE_MASK = 0,
```
Comment (esmirno) [inline src/plugins/intel_npu/src/plugin/npuw/attn/attn_subgraph.cpp]: if it is regular tile with mask - may be better to have TILE_WITH_MASK and REGULAR_TILE

File: src/plugins/intel_npu/src/plugin/npuw/attn/attn_subgraph.cpp [hunk 2]
```diff
@@ -275,25 +276,54 @@ void ensure_hfa_requests(ov::npuw::v1::subgraphs::InferContext& ctx, RuntimeStat
...
+    if (hfa->_compiled_tile_no_mask_model) {
+        state.hfa_requests.infer_requests[HFARequestSet::REGULAR_TILE_NO_MASK] =
+            hfa->_compiled_tile_no_mask_model->create_infer_request();
+    }
     state.hfa_requests.infer_requests[HFARequestSet::FINAL_TILE] = state.base_request;
     if (is_piped) {
-        state.hfa_requests.pipeline_requests[HFARequestSet::REGULAR_TILE] =
+        state.hfa_requests.pipeline_requests[HFARequestSet::REGULAR_TILE_MASK] =
             hfa->_compiled_tile_model->create_infer_request();
+        if (hfa->_compiled_tile_no_mask_model) {
```
Comment (esmirno) [inline src/plugins/intel_npu/src/plugin/npuw/attn/attn_subgraph.cpp]: this is  clear place to spot a problem - initial no_mask_model might refer missing a model, so better to have compiled_tile, compiled_tile_with_mask

File: src/plugins/intel_npu/src/plugin/npuw/host_flash_attention.cpp [hunk 3]
```diff
@@ -1018,6 +1039,18 @@ std::optional<HostFlashAttention> HostFlashAttention::from(const std::shared_ptr
     // ========================================================================
     HostFlashAttention hfa;
     hfa._tile_model = tile_model;
+    if (fused_flash_attention) {
```
Comment (esmirno) [inline src/plugins/intel_npu/src/plugin/npuw/host_flash_attention.cpp]: do we have any tests for HFA - i think @intelgaoxiong introduced one - interestingly how non of it is failing with such change - could you please add some tests that shows MASKS behavior

**Review by esmirno — COMMENTED** [review]
overall LGTM - but tests better to always create since otherwise not clear what this change is fixing.

File: src/plugins/intel_npu/src/plugin/npuw/attn/attn_subgraph.cpp (hunk 1 above)
Comment (intelgaoxiong) [inline src/plugins/intel_npu/src/plugin/npuw/attn/attn_subgraph.cpp]: So REGULAR_TILE_MASK is the 1st hfa tile, REGULAR_TILE_NO_MASK is the middle tiles and FINAL_TILE is the last tile.

I think they are used for different cases, right?
when short prompt is not a full tile, we will use `REGULAR_TILE_MASK`,
the last tile is usually not full, we will have to use `FINAL_TILE` with mask?

Please add comments why have middle tiles skip mask and how they are used for different cases.

File: src/plugins/intel_npu/src/plugin/npuw/attn/attn_subgraph.cpp [hunk 4]
```diff
@@ -275,25 +276,54 @@ void ensure_hfa_requests(ov::npuw::v1::subgraphs::InferContext& ctx, RuntimeStat
...
+                                                                                               pipeline_tensor);
+        }
+    }
+
+    // Share final tile model tensors with the no-mask tile request (only the non-mask inputs)
+    if (hfa->_compiled_tile_no_mask_model) {
+        const size_t num_no_mask_inputs = hfa->_compiled_tile_no_mask_model->inputs().size();
+        for (size_t input_idx = 0; input_idx < num_no_mask_inputs; ++input_idx) {
+            const auto no_mask_tile_input = hfa->_compiled_tile_no_mask_model->inputs()[input_idx];
+            const auto final_tile_input = hfa->_compiled_final_tile_model->inputs()[input_idx];
```
Comment (intelgaoxiong) [inline src/plugins/intel_npu/src/plugin/npuw/attn/attn_subgraph.cpp]: Is there any mechanism to ensure input_idx are the same port in `_compiled_tile_no_mask_model` and `_compiled_final_tile_model`?

Otherwise, you would get a failure with `set_tensor`

File: src/plugins/intel_npu/src/plugin/npuw/attn/attn_subgraph.cpp [hunk 5]
```diff
@@ -786,8 +830,15 @@ ov::npuw::v1::subgraphs::RuntimeBehaviorFactory make_runtime_factory() {
                         auto present_key_tensor = hfa_inputs.at(sdpa_in.present_key);
                         auto attention_mask_tensor = hfa_inputs.at(sdpa_in.attention_mask);
                         auto present_value_tensor = hfa_inputs.at(sdpa_in.present_value);
-                        auto& regular_tile_request = state.hfa_requests.infer_requests[HFARequestSet::REGULAR_TILE];
+                        // If the regular tile is not fully filled, need to use the mask
+                        const bool use_mask = (actual_kv_length + 1) % tile_size != 0;
```
Comment (intelgaoxiong) [inline src/plugins/intel_npu/src/plugin/npuw/attn/attn_subgraph.cpp]: The last tile in every chunk must have mask, right?
Image

`// If the regular tile is not fully filled, need to use the mask`
I think only final tile can be not fully filled, regular tiles are always full?

So we can only use mask in the final tile, the regular tile can always skip mask safely?

**Comment by dmatveev** [comment]
@kkoryun can you please update the PR according to feedback?

File: src/plugins/intel_npu/src/plugin/npuw/host_flash_attention.cpp [hunk 6]
```diff
@@ -1044,8 +1063,10 @@ std::optional<HostFlashAttention> HostFlashAttention::from(const std::shared_ptr
 
     // ========================================================================
     // Step 8: Build tile model parameter index mapping
+    // The first 6 input indices are identical in both models regular and final
+    // final_tile_model has mask_tile (index 6)
     // ========================================================================
-    build_tile_param_mapping(hfa, tile_model);
+    build_tile_param_mapping(hfa, final_tile_model);
```
Comment (intelgaoxiong) [inline src/plugins/intel_npu/src/plugin/npuw/host_flash_attention.cpp:1072]: This builds the parameter index mapping using final_tile_model, but the resulting mapping is also reused by regular_tile_model.
This is only correct if:
The first 6 input parameters share identical indices between regular_tile_model and final_tile_model
final_tile_model has mask_tile appended exclusively at the end (e.g., index 6)

Could you confirm this assumption is guaranteed?

Comment (kkoryun) [reply src/plugins/intel_npu/src/plugin/npuw/host_flash_attention.cpp:1072]: Yes right, now the regular model has 1 less input and it is a mask input which is presented in final tile model The first 6 inputs are the same for both model, the mask is 7 (with index = 6). I added this comment above in the code. 
I can add more comment in the inputs enum definition, for example. Or it can be redesigned differently to remove such a mismatch

Comment (intelgaoxiong) [reply src/plugins/intel_npu/src/plugin/npuw/host_flash_attention.cpp:1072]: OK, looks fine.
Thanks for the explanation.

**Review by intelgaoxiong — APPROVED** [review]
LGTM. Thanks a lot!
Could you please update the new logic of mask skipping in PR description? @kkoryun

File: src/plugins/intel_npu/src/plugin/npuw/attn/attn_subgraph.cpp [hunk 7]
```diff
@@ -1047,6 +1051,9 @@ ov::npuw::v1::subgraphs::RuntimeBehaviorFactory make_runtime_factory() {
                         int64_t mask_tile_offset = 0;
                         int64_t past_kv_tiles = num_tiles - 1;  // tiles driven from past blocks
 
+                        // For the fused hfa, the regular tile model has no mask input (6 inputs)
+                        const bool uses_mask = hfa_desc->_compiled_tile_model->inputs().size() > tile_in.mask;
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/attn/attn_subgraph.cpp:1055]: does it holds true for the last chunk (very bottom of the attention triangle)?
it is right-padded today? I believe it used to be left-padded before?

File: src/plugins/intel_npu/src/plugin/npuw/host_flash_attention.cpp [hunk 8]
```diff
@@ -585,8 +601,11 @@ static std::shared_ptr<ov::Model> create_hfa_tile_model(const ov::Shape& q_shape
     // Create input parameters
     auto inputs = create_hfa_tile_inputs(q_shape, input_dtype, mask_dtype, tile_size, kv_num_heads);
 
-    // Convert all inputs to f32
-    auto f32_nodes = convert_inputs_to_f32(inputs, mask_dtype, compute_dtype);
+    // Convert all inputs to f32.
+    // For the fused operation only the final tile uses a mask (regular tiles skip mask for performance)
+    // For the non-fused operation all tiles require mask conversion
+    const bool use_mask = is_final_tile || !fused_flash_attention;
+    auto f32_nodes = convert_inputs_to_f32(inputs, mask_dtype, compute_dtype, use_mask);
```
[the remark this answers is not in the document]
Comment (dmatveev) [reply src/plugins/intel_npu/src/plugin/npuw/host_flash_attention.cpp:608]: I believe this time I agree with that machine says

**Review by dmatveev — CHANGES_REQUESTED** [review]
I think this code is written in assumption our mask is always defined by the sequence length.

In fact, it is not the case. Attention mask tiles used for tile computations (be it HFA or pyramid) are taken from the 2D in-graph attention mask that is computed by the model architecture and takes into account:
1. position_ids
2. a 1D attention mask passed outside
3. a sliding window, if any
4. a local attention, if any.

It seems for the local attention this new simplified masking mechanism may not work correctly. The generic case with user-specific mask may also be broken - I think we need to take a closer look on this.

File: src/plugins/intel_npu/src/plugin/npuw/attn/attn_subgraph.cpp (hunk 7 above)
Comment (kkoryun) [reply src/plugins/intel_npu/src/plugin/npuw/attn/attn_subgraph.cpp:1055]: In the case of a _fused_ HFA, the _final_ model has a mask input, when any _regular_ model has no such input (including very bottom of the attention triangle)
In the case of a _non-fused_ HFA, both subgraphs have mask input.

**Review by dmatveev — APPROVED** [review]
Approved with comments - works for causal mask only, not applicable for the generic case, can't be enabled by default in the current form.
